### PR TITLE
PL-1497 fetch circleci build artifacts

### PR DESCRIPTION
**Pull request by sumanmukherjee03**
This code adds functionality to download build artifact from circleci using bub

Assumptions:
1. bub assumes that you are in the repo dir you need to run the command for
2. The new code reuses the code from ```bub check``` which assumes that you want the artifact of the latest build from current HEAD

Usage is like so :
```
bub circle artifact -h # Shows help menu
bub circle artifact -p $HOME junit.xml # Downloads the junit.xml artifact to $HOME dir
bub circle artifact -p $HOME/dir-thats-not-there junit.xml # Throws error and exits with exit code 1
bub circle artifact junit.xml # Downloads the junit.xml file to current dir
bub circle artifact artifact-thats-not-there.xml # NOOP : Doesnt do anything
```

### CODE REVIEW

File: cmd/other_cmd.go [hunk 1]
```diff
@@ -57,6 +61,39 @@ func buildCircleCmds(cfg *core.Configuration, manifest *core.Manifest) []cli.Com
...
+					return fmt.Errorf("You have too many arguments. The only argument accepted is the name of the artifact you want to download.")
+				}
+				name := c.Args().Get(0)
+				dir := c.String("path")
+				_, err := os.Stat(dir)
+				if len(dir) > 0 && err != nil {
+					return err
+				}
+				if len(dir) == 0 {
+					dir, err = filepath.Abs(filepath.Dir(os.Args[0]))
```
Comment (philwhln) [inline cmd/other_cmd.go]: This is getting `dir` for `os.Args[0]`, but `os.Args[0]` is the path to `bub` (e.g. `/usr/bin/bin`).
Should this be `pwd` instead?

File: cmd/other_cmd.go [hunk 2]
```diff
@@ -57,6 +61,39 @@ func buildCircleCmds(cfg *core.Configuration, manifest *core.Manifest) []cli.Com
...
+			Action: func(c *cli.Context) error {
+				if c.NArg() < 1 {
+					return fmt.Errorf("You have too few arguments. Please provide the name of the artifact you want to download.")
+				}
+				if c.NArg() > 1 {
+					return fmt.Errorf("You have too many arguments. The only argument accepted is the name of the artifact you want to download.")
+				}
+				name := c.Args().Get(0)
+				dir := c.String("path")
+				_, err := os.Stat(dir)
```
Comment (philwhln) [inline cmd/other_cmd.go]: Move to line 86 after you've got the passed arg for dir of current directory?

File: cmd/other_cmd.go [hunk 3]
```diff
@@ -57,6 +61,39 @@ func buildCircleCmds(cfg *core.Configuration, manifest *core.Manifest) []cli.Com
...
+				if len(dir) > 0 && err != nil {
```
Comment (philwhln) [inline cmd/other_cmd.go]: Should just be checking for err.

File: integrations/ci/circle.go [hunk 4]
```diff
@@ -143,3 +124,82 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
 	}
 	return nil, NoBuildFound
 }
+
+func (c *Circle) GetArtifact(m *core.Manifest, n, p string) error {
```
Comment (philwhln) [inline integrations/ci/circle.go]: Can we give fuller names for `n` and `p`?
`m` is ok, because it's obviously a `Manifest`, but can't tell what the others are.

File: integrations/ci/circle.go [hunk 5]
```diff
@@ -143,3 +124,82 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
...
+		log.Printf("Artifact URL: %s", a.URL)
+	}
+}
+
+func downloadArtifact(a *circleci.Artifact, path, token string) {
+	if len(a.URL) > 0 {
+		log.Printf("Download Path: %s", path)
+		err := utils.DownloadFile(path, fmt.Sprintf("%s?circle-token=%s", a.URL, token))
+		if err != nil {
+			panic(err)
```
Comment (philwhln) [inline integrations/ci/circle.go]: Is it common to `panic` in the `bub` code? Usually `panic` is a last resort. `return err` would be better, so caller can decide if they want to abort or not.

File: integrations/ci/circle.go [hunk 6]
```diff
@@ -143,3 +124,82 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
...
+	if err != nil {
+		return b, err
+	}
+	return b, nil
+}
+
+func getBuildArtifacts(c *circleci.Client, account, repo string, b *circleci.Build) ([]*circleci.Artifact, error) {
+	var arr []*circleci.Artifact
+	arr, err := c.ListBuildArtifacts(account, repo, b.BuildNum)
+	if err != nil {
```
Comment (philwhln) [inline integrations/ci/circle.go]: `if` statement is redundant if you're doing `return arr, err` regardless.

File: integrations/ci/circle.go [hunk 7]
```diff
@@ -143,3 +124,82 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
...
+
+	arr, err := getBuildArtifacts(c.client, c.GetAccount(), m.Repository, b)
+	if err != nil {
+		return err
+	}
+
+	for _, a := range arr {
+		if strings.Contains(a.URL, n) {
+			defer logArtifactDetails(a)
+			defer downloadArtifact(a, p, c.client.Token)
```
Comment (philwhln) [inline integrations/ci/circle.go]: Is the purpose of `defer` here for making them async? Does logging the artifact url need to be done in async to `downloadArtifact`?

File: integrations/ci/circle.go (hunk 6 above)
Comment (sumanmukherjee03) [reply integrations/ci/circle.go]: There was an error. The success case on 190 now returns arr, nil.

File: cmd/other_cmd.go (hunk 1 above)
Comment (sumanmukherjee03) [reply cmd/other_cmd.go]: It was resolving to the name of the binary instead of the path to the binary. Fixed it with ```os.Getwd()``` instead.

File: cmd/other_cmd.go (hunk 2 above)
Comment (sumanmukherjee03) [reply cmd/other_cmd.go]: Logic is fixed now.

File: cmd/other_cmd.go (hunk 3 above)
Comment (sumanmukherjee03) [reply cmd/other_cmd.go]: Also fixed.

File: integrations/ci/circle.go (hunk 4 above)
Comment (sumanmukherjee03) [reply integrations/ci/circle.go]: Updated

File: integrations/ci/circle.go (hunk 5 above)
Comment (sumanmukherjee03) [reply integrations/ci/circle.go]: There is 1 panic in the code. I have removed the panic and used ```log.Fatal``` instead.

File: integrations/ci/circle.go (hunk 7 above)
Comment (sumanmukherjee03) [reply integrations/ci/circle.go]: Moved the logging to the caller fn in other_cmd.go. Also converted the GetArtifact function to really return a circleci artifact object or an error. So, now we have just 1 defer call to download the artifact. The downloading seems like a side effect, hence the defer.

File: integrations/ci/circle.go [hunk 8]
```diff
@@ -143,3 +124,73 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
...
+	}
+
+	for _, a := range arr {
+		if strings.Contains(a.URL, fname) {
+			artifact = a
+			defer func(a *circleci.Artifact, path, token string) {
+				if len(a.URL) > 0 {
+					err := utils.DownloadFile(path, fmt.Sprintf("%s?circle-token=%s", a.URL, token))
+					if err != nil {
+						retErr = err
```
Comment (philwhln) [inline integrations/ci/circle.go]: Hard to follow this behavior, since assigning to the named return arg `retErr` and `defer`ing, but also calling `return artifact, nil` at the end of the function, so not clear what gets returned.

Possible to do this without the `defer` to keep it simple?

File: integrations/ci/circle.go [hunk 9]
```diff
@@ -91,42 +96,18 @@ func configurationExist() (bool, error) {
 }
 
 func (c *Circle) CheckBuildStatus(m *core.Manifest) error {
```
Comment (philwhln) [inline integrations/ci/circle.go]: Discussed IRL. Pretty much revert this to how it was but split out into `GetCompletedBuild` and use that to call `isSuccess(b)` or get artifacts for the completed build.

**Comment by sumanmukherjee03** [comment]
@philwhln i think i have addressed all the comments. Could you please take another look. Thanks :) .

File: integrations/ci/circle.go [hunk 10]
```diff
@@ -143,3 +155,38 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
...
+	}
+
+	arr, err := getBuildArtifacts(c.client, c.GetAccount(), m.Repository, b)
+	if err != nil {
+		return err
+	}
+
+	for _, a := range arr {
+		if strings.Contains(a.URL, fname) {
+			if len(a.URL) > 0 {
```
Comment (philwhln) [inline integrations/ci/circle.go]: This seems redundant, unless `fname` was also `""`, since you checked `strings.Contains(a.URL, fname)` previously.

File: integrations/ci/circle.go [hunk 11]
```diff
@@ -143,3 +155,38 @@ func (c *Circle) checkBuildStatus(head string, m *core.Manifest) (*circleci.Buil
...
+				err := utils.DownloadFile(path, fmt.Sprintf("%s?circle-token=%s", a.URL, c.client.Token))
```
Comment (philwhln) [inline integrations/ci/circle.go]: Could just be
`return utils.DownloadFile(path, fmt.Sprintf("%s?circle-token=%s", a.URL, c.client.Token))`
?
